tests/platform/psu_controller.py

```diff
@@ -0,0 +1,113 @@
+"""
+Fixture for controlling PSUs of DUT
+
+This file defines fixture psu_controller which is for controlling PSUs of DUT. The fixture uses factory design pattern
+and returns a function for creating PSU controller instance. The function takes two arguments:
+* hostname - hostname of the DUT
+* asic_type - asic type of the DUT
+Based on these two inputs, different PSU controller implemented by different vendors could be returned.
+
+The PSU controller implemented by each vendor must be a subclass of the PsuControllerBase class and implements the
+methods defined in the base class.
```

Comment:
@wangxin: Can you explain how and where vendors should implement this base class?

Reply:
Hi Joe,

This file just contains the base class definition and psu_controller fixture. Mellanox specific implementation is not in this file. 

Each vendor can implement their own PSU controller by calling APIs of the PSU controlling devices used in their testbed. The requiremenet is that the PSU controller sub-class must implement the methods defined in the base class. Expected return values are specified in docstring of each base class's methods. Vendor specific PSU controller implementation script should be put under vendor's subfolder. 

For example, Mellanox specific implementation is in file test/platform/mellanox/mellanox_psu_controller.py. The implementation uses Mellanox internal lab configuration database to find out IP address of PSU controller. Then it uses SNMP to get information for the PSU controlling device and control power on/off to PSUs of SONiC switch. Please note that because the mellanox_psu_controller.py file contains Mellanox internal lab configuration data, it is not included in this PR.

Reply:
> This file just contains the base class definition and psu_controller fixture. Mellanox specific implementation is not in this file.
> Please note that because the mellanox_psu_controller.py file contains Mellanox internal lab configuration data, it is not included in this PR.

However, mellanox_psu_controller.py is referenced in the `_make_psu_controller()` function of this file. Why?

Reply:
The psu_controller fixture uses factory pattern to create psu controller object based on SONiC switch's asic_type and hostname. The _make_psu_controller() function is returned to testing scripts (by `yield _make_psu_controller`). When it is called in testing script with argument 'hostname' and 'asic_type', it can return an initialized psu controller object. Then the testing script can get the psu controller object for controlling power off/on to PSUs of SONiC switch.

Because vendor specific PSU controller is not implemented in test/platform/psu_controller.py, the _make_psu_controller function need to import vendor specific implementation for creating the actual PSU controller object based on vendor information passed in by the asic_type argument. So, if asic_type is "mellanox", the mellanox_psu_controller would be referenced. If asic_type is not "mellanox", that code branch referencing mellanox_psu_controller would not be executed.

Reply:
Understood. However, I'm not keen on a checked-in file referencing a file that is not checked in.

Also, to clarify, the "PSU controller" controls a PDU which the switch is plugged into, correct? If so, why is the "PSU controller" dictated by the hostname and ASIC type?

Reply:
How about I replace actual testbed configuration with dummy config in test/platform/mellanox/mellanox_psu_controller.py and include it in this PR? It can be served as a reference PSU controller implementation.

Correct, "PSU controller" also can be the PDU device itself. In Mellanox lab, we use a type of PDU device which have multiple power outlets. Each power outlet can be connected to a switch PSU and provides electricty power. The PDU devices have IP addresses and provides multiple management interfaces for setting and getting power outlets configuration information, status, etc.

Usually two PSUs of a SONiC switch are connected to two power outlets of a PDU device. We have an internal "database" stores the mapping between SONiC switches and IP address of the PDU devices that they are connected to. That's why hostname information is required for creating PSU controller object. I believe other vendor also needs this information for querying information of PDU device that the DUT switch is connected to.

Different vendor may use different types of PDU device and needs different PSU controller implementation. The asic_type patameter is for importing vendor specific PSU controller implementation.

Reply:
I think it would be better to include a dummy controller file, rather than to reference a file which isn't checked in.

Reply:
Sure. Just added a commit contains the mellanox_psu_controller implementation.